gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala

```diff
@@ -674,62 +679,48 @@ case class TransformPostOverrides(isAdaptiveContext: Boolean) extends Rule[Spark
-        isGlutenTableCacheInternal(i)
-      case q: QueryStageExec =>
-        // Compatible with Spark3.5 `TableCacheQueryStage`
-        isGlutenTableCache(q.plan)
-      case _ => false
-    }
-  }
-}
-
 // This rule will try to add RowToColumnarExecBase and ColumnarToRowExec
-// to support vanilla columnar scan.
+// to support vanilla columnar operators.
 case class VanillaColumnarPlanOverrides(session: SparkSession) extends Rule[SparkPlan] {
   @transient private val planChangeLogger = new PlanChangeLogger[SparkPlan]()
 
   private def replaceWithVanillaColumnarToRow(plan: SparkPlan): SparkPlan = plan match {
-    case c2r: ColumnarToRowExecBase if isVanillaColumnarReader(c2r.child) =>
-      ColumnarToRowExec(c2r.child)
-    case c2r: ColumnarToRowExec if isVanillaColumnarReader(c2r.child) =>
-      c2r
-    case _ if isVanillaColumnarReader(plan) =>
-      BackendsApiManager.getSparkPlanExecApiInstance.genRowToColumnarExec(ColumnarToRowExec(plan))
+    case _ if PlanUtil.isGlutenColumnarOp(plan) =>
+      plan.withNewChildren(plan.children.map {
+        c =>
+          val child = replaceWithVanillaColumnarToRow(c)
+          if (PlanUtil.isVanillaColumnarOp(child)) {
+            BackendsApiManager.getSparkPlanExecApiInstance.genRowToColumnarExec(
+              ColumnarToRowExec(child))
+          } else {
+            child
+          }
+      })
     case _ =>
       plan.withNewChildren(plan.children.map(replaceWithVanillaColumnarToRow))
   }
 
-  private def isVanillaColumnarReader(plan: SparkPlan): Boolean = plan match {
-    case _: BatchScanExec | _: DataSourceScanExec =>
-      !plan.isInstanceOf[GlutenPlan] && plan.supportsColumnar
-    case i: InMemoryTableScanExec =>
-      if (InMemoryTableScanHelper.isGlutenTableCache(i)) {
-        // `InMemoryTableScanExec` do not need extra RowToColumnar or ColumnarToRow
-        false
-      } else {
-        !plan.isInstanceOf[GlutenPlan] && plan.supportsColumnar
-      }
-    case _ => false
+  private def replaceWithVanillaRowToColumnar(plan: SparkPlan): SparkPlan = plan match {
+    case _ if PlanUtil.isVanillaColumnarOp(plan) =>
+      plan.withNewChildren(plan.children.map {
+        c =>
+          val child = replaceWithVanillaRowToColumnar(c)
+          if (PlanUtil.isGlutenColumnarOp(child)) {
+            RowToColumnarExec(
+              BackendsApiManager.getSparkPlanExecApiInstance.genColumnarToRowExec(child))
+          } else {
+            child
+          }
+      })
+    case _ =>
+      plan.withNewChildren(plan.children.map(replaceWithVanillaRowToColumnar))
   }
 
-  def apply(plan: SparkPlan): SparkPlan =
-    if (GlutenConfig.getConf.enableVanillaVectorizedReaders) {
-      val newPlan = replaceWithVanillaColumnarToRow(plan)
-      planChangeLogger.logRule(ruleName, plan, newPlan)
-      newPlan
-    } else {
-      plan
-    }
+  def apply(plan: SparkPlan): SparkPlan = {
+    val newPlan = replaceWithVanillaRowToColumnar(replaceWithVanillaColumnarToRow(plan))
```

Comment:
Can we add config `enableVanillaVectorizedReaders` back ?

Reply:
I think here this config check is not needed. If this conf is false, then scan ops' `supportsColumnar` should be false, then the logic here would not change the plan. Please let me know if anything I missed.

Reply:
make sense